sqlair.go

```diff
@@ -0,0 +1,230 @@
+package sqlair
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"reflect"
+
+	"github.com/canonical/sqlair/internal/expr"
+)
+
+// Statement represents a SQL statemnt with valid SQLair expressions.
+// It is ready to be run on a SQLair DB.
+type Statement struct {
+	pe *expr.PreparedExpr
+}
+
+// Prepare expands the types mentioned in the SQLair expressions and checks
+// the SQLair parts of the query are well formed.
+// typeInstantiations must contain an instance of every type mentioned in the
+// SQLair expressions of the query. These are used only for type information.
+func Prepare(query string, typeInstantiations ...any) (*Statement, error) {
+	parser := expr.NewParser()
+	parsedExpr, err := parser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+	preparedExpr, err := parsedExpr.Prepare(typeInstantiations...)
+	if err != nil {
+		return nil, err
+	}
+	return &Statement{pe: preparedExpr}, nil
+}
+
+// MustPrepare is the same as prepare except that it panics on error.
+func MustPrepare(query string, typeInstantiations ...any) *Statement {
+	s, err := Prepare(query, typeInstantiations...)
+	if err != nil {
+		panic(err)
+	}
+	return s
+}
+
+type DB struct {
+	db *sql.DB
+}
+
+func NewDB(db *sql.DB) *DB {
+	return &DB{db: db}
+}
+
+// sqlDB returns the underlying database object.
+func (db *DB) sqlDB() *sql.DB {
+	return db.db
+}
+
+// Exec executes an SQLair Statement without returning any rows.
+// Exec uses ExecContext with context.Background internally.
+func (db *DB) Exec(s *Statement, inputStructs ...any) (sql.Result, error) {
```

Comment:
As we talked, can we please leave the Exec functions out for now?  We may end up doing them in terms of the Query API itself too, so let's focus on makign that one really nice and then we figure how to do the one shot calls.

Reply:
Removed the Exec function.